source/grid/grid_tools_nontemplates.cc

```diff
@@ -130,126 +130,139 @@ namespace GridTools
 
         return (1.0 / 6.0) * std::abs((a - d) * cross_product_3d(b - d, c - d));
       }
+    else if (vertex_indices.size() == 5) // pyramid
+      {
+        // This remarkably simply formula comes from Equation 4 of
+        // "Calculation of the volume of a general hexahedron for flow
+        // predictions", Davies and Salmond, AIAA Journal vol. 23 no. 6.
+        const auto &x0 = all_vertices[vertex_indices[0]];
+        const auto &x1 = all_vertices[vertex_indices[1]];
+        const auto &x2 = all_vertices[vertex_indices[2]];
+        const auto &x3 = all_vertices[vertex_indices[3]];
+        const auto &x4 = all_vertices[vertex_indices[4]];
+
+        auto v01 = x1 - x0;
+        auto v02 = x2 - x0;
+        auto v03 = x3 - x0;
+        auto v04 = x4 - x0;
+        auto v21 = x2 - x1;
+
+        // doing high - low consistently puts us of by -1 from the original
+        // paper in the first term
+        return -1.0 * v04 * cross_product_3d(v21, v03) / 6.0 +
+               v03 * cross_product_3d(v01, v02) / 12.0;
+      }
```

Comment:
This is a curious formula whose origin I don't understand. Is it only true for pyramids for which the four points of the base are in a single plane? In that case, the cross-product would have something to do with the area of the base, and the dot product with v04 would be like multiplying the base area by something relating to the height of the pyramid. But I don't understand the origin of the second term. It only contains points of the base. Is it a correct for the nonplanarity of the base? (The second term is zero for a planar base.)

Reply:
It is derived in the paper I mention in the comment - I don't have a good intuition for it either. It works for non-planar bases, which is what I picked for the 'general pyramid' case:

https://github.com/dealii/dealii/blob/6ec9ecb39d67d5a073295fc3868880669f359943/tests/simplex/cell_measure_01.cc#L208-L213

the first four points are definitely nonplanar.